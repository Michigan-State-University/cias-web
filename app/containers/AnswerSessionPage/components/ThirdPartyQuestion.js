import React, { useState, useEffect } from 'react';
import PropTypes from 'prop-types';

import ThirdPartyQuestionLayout from '../layouts/ThirdPartyQuestionLayout';

const ThirdPartyQuestion = ({
  question,
  answerBody,
  selectAnswer,
  isMobile,
  disabled,
}) => {
  const [selectedAnswerIndex, setSelectedAnswerIndex] = useState(null);
  const {
    body: {
      data,
      variable: { name },
    },
    id,
  } = question;

  useEffect(() => {
    setSelectedAnswerIndex(answerBody.length ? answerBody[0].index : null);
  }, [id]);

  const handleClick = (value, reportTemplateIds, numericValue, index) => {
    selectAnswer([
      {
        var: name,
        value,
        report_template_ids: reportTemplateIds,
        numeric_value: numericValue,
        index,
      },
    ]);
    setSelectedAnswerIndex(index);
  };

  return (
    <ThirdPartyQuestionLayout
      data={data}
      handleClick={handleClick}
      questionId={id}
      selectedAnswerIndex={selectedAnswerIndex}
      isMobile={isMobile}
      disabled={disabled}
    />
  );
};

ThirdPartyQuestion.propTypes = {
  question: PropTypes.object.isRequired,
  answerBody: PropTypes.any,
  selectAnswer: PropTypes.func,
  isMobile: PropTypes.bool,
  disabled: PropTypes.bool,
};

export default ThirdPartyQuestion;
