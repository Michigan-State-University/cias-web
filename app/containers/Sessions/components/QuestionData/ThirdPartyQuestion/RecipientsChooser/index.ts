export * from './RecipientsChooser';
export * from './types';
