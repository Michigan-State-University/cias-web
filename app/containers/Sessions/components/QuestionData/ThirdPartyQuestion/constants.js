export const ADD = 'app/EditSessionPage/ThirdParty/ADD';
export const UPDATE_ANSWER = 'app/EditSessionPage/ThirdParty/UPDATE';
export const REMOVE = 'app/EditSessionPage/ThirdParty/REMOVE';
export const REORDER = 'app/EditSessionPage/ThirdParty/REORDER';
