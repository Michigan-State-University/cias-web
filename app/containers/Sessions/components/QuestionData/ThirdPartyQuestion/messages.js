import { defineMessages } from 'react-intl';

export const scope = 'app.containers.ThirdParty';

export default defineMessages({
  placeholder: {
    id: `${scope}.placeholder`,
    defaultMessage: 'Third Party {index}',
  },
  addAnswer: {
    id: `${scope}.addAnswer`,
    defaultMessage: 'Add new Third Party',
  },
  emailError: {
    id: `${scope}.emailError`,
    defaultMessage: 'Incorrect email address entered',
  },
});
