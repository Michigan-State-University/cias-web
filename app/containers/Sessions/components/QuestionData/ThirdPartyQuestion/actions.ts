import { updateQuestionData } from 'global/reducers/questions';
import { ThirdPartyReportQuestionData } from 'models/Question';

import { REORDER } from './constants';

export const reorderAnswersAction = (items: ThirdPartyReportQuestionData[]) =>
  updateQuestionData({ type: REORDER, data: { items } });
