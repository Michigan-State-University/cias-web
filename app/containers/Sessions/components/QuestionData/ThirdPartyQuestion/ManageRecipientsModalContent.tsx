import React from 'react';
import { useIntl } from 'react-intl';

import RedBin from 'assets/svg/bin-red-no-bg.svg';

import Tabs from 'components/Tabs';
import { ModalContentRenderer } from 'components/Modal';
import { StripedTR, Table, TBody, TD } from 'components/Table';
import Row from 'components/Row';
import Text from 'components/Text';
import { ImageButton } from 'components/Button';

import messages from './messages';
import { Recipients } from './types';

export type Props = {};

export const ManageRecipientsModalContent: ModalContentRenderer<Recipients> = ({
  modalState,
}) => {
  const { formatMessage } = useIntl();

  if (!modalState) return null;
  const { emails } = modalState;

  return (
    // @ts-ignore
    <Tabs
      mt={14}
      withBottomBorder
      emphasizeActiveLink
      containerProps={{ mb: 0, mt: 16 }}
    >
      {/* @ts-ignore */}
      <div label={formatMessage(messages.emailRecipients)}>
        <Table width="100%" tableLayout="fixed">
          <TBody>
            {emails.map((email) => (
              <StripedTR height="auto">
                <TD padding="12px 8px">
                  <Row justify="between" align="center" gap={8}>
                    <Text fontSize={15} lineHeight={1.5} truncate>
                      {email}
                    </Text>
                    <ImageButton
                      src={RedBin}
                      onClick={() => {}}
                      title={formatMessage(messages.deleteRecipient)}
                    />
                  </Row>
                </TD>
              </StripedTR>
            ))}
          </TBody>
        </Table>
      </div>
      {/* @ts-ignore */}
      <div label={formatMessage(messages.faxRecipients)}></div>
    </Tabs>
  );
};
