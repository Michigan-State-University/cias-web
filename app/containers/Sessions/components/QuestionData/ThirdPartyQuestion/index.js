import React, { useCallback, useState } from 'react';
import PropTypes from 'prop-types';
import { compose } from 'redux';
import { connect } from 'react-redux';
import { createStructuredSelector } from 'reselect';
import { injectIntl } from 'react-intl';

import {
  makeSelectSelectedQuestion,
  updateQuestionData,
} from 'global/reducers/questions';
import { themeColors } from 'theme';

import Box from 'components/Box';
import Column from 'components/Column';
import HoverableBox from 'components/Box/HoverableBox';
import PlusCircle from 'components/Circle/PlusCircle';
import Row from 'components/Row';
import Text from 'components/Text';
import { DndSortable } from 'components/DragAndDrop';

import messages from './messages';
import { ADD, UPDATE_ANSWER, REMOVE } from './constants';
import { reorderAnswersAction } from './actions';
import { ThirdPartyQuestionAnswer } from './ThirdPartyQuestionAnswer';

const ThirdPartyQuestion = ({
  selectedQuestion,
  addAnswer,
  updateAnswer,
  removeAnswer,
  reorderAnswers,
  isNarratorTab,
  editingPossible,
  intl: { formatMessage },
}) => {
  const [hoveredIndex, setHoveredIndexIndex] = useState(-1);
  const handleMouseLeave = useCallback(
    () => setHoveredIndexIndex(-1),
    [setHoveredIndexIndex],
  );

  const {
    id,
    body: { data },
  } = selectedQuestion;

  const canDeleteAnswer = data.length > 1;

  const isNarratorTabOrEditNotPossible = isNarratorTab || !editingPossible;

  const handleChange = useCallback(
    (index, answer) => {
      updateAnswer(index, answer);
    },
    [updateAnswer],
  );

  const onDragEnd = (_, items, hasChanged) => {
    if (!hasChanged) return;

    reorderAnswers(items);
  };

  return (
    <Column mt={10}>
      <DndSortable onDragEnd={onDragEnd} items={data} selector={null}>
        {({ item, index, dragHandleProps }) => (
          <ThirdPartyQuestionAnswer
            key={`question-${id}-answer-${index}-item`}
            questionId={id}
            answer={item}
            isNarratorTabOrEditNotPossible={isNarratorTabOrEditNotPossible}
            isNarratorTab={isNarratorTab}
            editingPossible={editingPossible}
            index={index}
            dragHandleProps={dragHandleProps}
            onMouseEnter={setHoveredIndexIndex}
            onMouseLeave={handleMouseLeave}
            hoveredIndex={hoveredIndex}
            canDeleteAnswer={canDeleteAnswer}
            onChange={handleChange}
            onRemove={removeAnswer}
          />
        )}
      </DndSortable>
      <Row display="flex" hidden={isNarratorTabOrEditNotPossible}>
        <HoverableBox px={21} py={14} onClick={addAnswer}>
          <Box>
            <Row align="center">
              <PlusCircle mr={12} />
              <Text fontWeight="bold" color={themeColors.secondary}>
                {formatMessage(messages.addAnswer)}
              </Text>
            </Row>
          </Box>
        </HoverableBox>
      </Row>
    </Column>
  );
};

ThirdPartyQuestion.propTypes = {
  selectedQuestion: PropTypes.object.isRequired,
  intl: PropTypes.object.isRequired,
  addAnswer: PropTypes.func.isRequired,
  updateAnswer: PropTypes.func.isRequired,
  removeAnswer: PropTypes.func.isRequired,
  reorderAnswers: PropTypes.func.isRequired,
  isNarratorTab: PropTypes.bool,
  editingPossible: PropTypes.bool,
};

const mapStateToProps = createStructuredSelector({
  selectedQuestion: makeSelectSelectedQuestion(),
});

const mapDispatchToProps = {
  addAnswer: () => updateQuestionData({ type: ADD }),
  updateAnswer: (index, value) =>
    updateQuestionData({ type: UPDATE_ANSWER, data: { index, value } }),
  removeAnswer: (index) =>
    updateQuestionData({ type: REMOVE, data: { index } }),
  reorderAnswers: reorderAnswersAction,
};

const withConnect = connect(mapStateToProps, mapDispatchToProps);

export default injectIntl(compose(withConnect)(ThirdPartyQuestion));
