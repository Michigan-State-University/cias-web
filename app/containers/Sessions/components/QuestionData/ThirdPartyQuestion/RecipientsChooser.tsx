import React, { FC, useMemo } from 'react';
import { useIntl } from 'react-intl';

import { themeColors } from 'theme';

import { htmlToPlainText } from 'utils/htmlToPlainText';

import Row from 'components/Row';
import Column from 'components/Column';
import { TextButton } from 'components/Button';
import { ModalProps, ModalType, useModal } from 'components/Modal';

import messages from './messages';
import { divideRecipients } from './utils';
import { RecipientList } from './RecipientList';
import { MANAGE_RECIPIENTS_MODAL_WIDTH } from './constants';
import { ManageRecipientsModalContent } from './ManageRecipientsModalContent';
import { Recipients } from './types';

export type Props = {
  hidden: boolean;
  disabled: boolean;
  recipients: string;
  modalTitle: string;
};

export const RecipientsChooser: FC<Props> = ({
  hidden,
  disabled,
  recipients,
  modalTitle,
}) => {
  const { formatMessage } = useIntl();

  const dividedRecipients = useMemo(
    () => divideRecipients(recipients),
    [recipients],
  );

  const modalProps = useMemo<ModalProps['props']>(
    () => ({
      title:
        htmlToPlainText(modalTitle) ||
        formatMessage(messages.manageRecipientsModalDefaultTitle),
      description: formatMessage(messages.manageRecipientsModalDescription),
      width: MANAGE_RECIPIENTS_MODAL_WIDTH,
    }),
    [modalTitle],
  );

  const { openModal, Modal } = useModal<Recipients>({
    type: ModalType.Modal,
    modalContentRenderer: ManageRecipientsModalContent,
    props: modalProps,
  });

  return (
    <>
      <Modal />
      <Column mb={10} ml={40} width="auto" hidden={hidden}>
        <RecipientList
          label={formatMessage(messages.emailRecipients)}
          noItemsLabel={formatMessage(messages.noEmailRecipients)}
          recipients={dividedRecipients.emails}
        />
        <RecipientList
          label={formatMessage(messages.faxRecipients)}
          noItemsLabel={formatMessage(messages.noFaxRecipients)}
          recipients={dividedRecipients.faxes}
          mt={8}
        />
        <Row mt={12}>
          <TextButton
            disabled={disabled}
            outlined
            buttonProps={{
              color: themeColors.secondary,
            }}
            onClick={() => openModal(dividedRecipients)}
          >
            {formatMessage(messages.manageRecipients)}
          </TextButton>
        </Row>
      </Column>
    </>
  );
};
