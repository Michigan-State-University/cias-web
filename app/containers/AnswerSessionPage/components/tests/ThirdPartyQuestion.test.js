/**
 *
 * Tests for ThirdPartyQuestion
 *
 */

import React from 'react';
import { render } from '@testing-library/react';
import { IntlProvider } from 'react-intl';
import { MemoryRouter } from 'react-router-dom';
import { DEFAULT_LOCALE } from 'i18n';

import ThirdPartyQuestion from '../ThirdPartyQuestion';

describe('<ThirdPartyQuestion />', () => {
  const mockedFunctions = {
    selectAnswer: jest.fn(),
  };

  const defaultProps = {
    question: {
      body: {
        data: [
          {
            payload: 'test',
            value: 'example1@mail.com',
          },
          {
            payload: 'test',
            value: 'example2@mail.com',
          },
        ],
        variable: '1',
      },
      id: 'test-id',
      settings: {
        proceed_button: true,
      },
    },
    questionIndex: 0,
    answerBody: [],
    ...mockedFunctions,
  };

  it('Expect to not log errors in console', () => {
    const spy = jest.spyOn(global.console, 'error');
    render(
      <IntlProvider locale={DEFAULT_LOCALE}>
        <MemoryRouter>
          <ThirdPartyQuestion {...defaultProps} />
        </MemoryRouter>
      </IntlProvider>,
    );
    expect(spy).not.toHaveBeenCalled();
  });

  it('Should render and match the snapshot', () => {
    const { container } = render(
      <IntlProvider locale={DEFAULT_LOCALE}>
        <MemoryRouter>
          <ThirdPartyQuestion {...defaultProps} />
        </MemoryRouter>
      </IntlProvider>,
    );
    expect(container).toMatchSnapshot();
  });
});
