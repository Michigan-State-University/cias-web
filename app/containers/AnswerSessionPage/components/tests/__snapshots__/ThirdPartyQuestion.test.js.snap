// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ThirdPartyQuestion /> Should render and match the snapshot 1`] = `
<div>
  <div
    class="Column-sc-4zswzq-0 ekpwZH"
  >
    <div
      class="Box-sc-1vhutan-0 dvfkZg"
    >
      <div
        class="Row-y1cs0u-0 kdnhbN"
      >
        <div
          class="Box-sc-1vhutan-0 HoverableBox-zvip6a-0 kmyFFE"
          width="calc(100% + 21px)"
        >
          <div
            class="Row-y1cs0u-0 hBrQLN"
            height="44"
          >
            <img
              class="Img-sc-55j3su-0 iYjftT"
              data-cy="single-question-0-checkbox"
              src="IMAGE_MOCK"
            />
            <div>
              test
            </div>
          </div>
        </div>
      </div>
      <div
        class="Row-y1cs0u-0 kdnhbN"
      >
        <div
          class="Box-sc-1vhutan-0 HoverableBox-zvip6a-0 kmyFFE"
          width="calc(100% + 21px)"
        >
          <div
            class="Row-y1cs0u-0 hBrQLN"
            height="44"
          >
            <img
              class="Img-sc-55j3su-0 iYjftT"
              data-cy="single-question-1-checkbox"
              src="IMAGE_MOCK"
            />
            <div>
              test
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
